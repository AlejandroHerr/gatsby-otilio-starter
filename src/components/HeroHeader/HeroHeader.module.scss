@import '../../colors';
@import '../../fonts';

.hero_header {
  position: relative;
  padding-top: 6em;
}

.hero_header__inner {
  max-width: 32em;
  margin: 0 auto;
  padding: 0 2em;
}

.hero_header--has_image {
  padding: 16em 0 2em;
  background: $color-black;
  text-shadow: 0 1px 2px rgba(0, 0, 0, .1);
  overflow: hidden;
}

.hero_image {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: no-repeat scroll center center / cover $color-black;
  animation: .5s ease-in forwards fadein;
  z-index: 5;

  img {
    margin: 0;
  }
}

.hero_image__shadow {
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 12em;
  background:
    linear-gradient(
      to bottom,
      rgba(0, 0, 0, 0) 0%,
      rgba(0, 0, 0, .5) 100%
    );
  z-index: 8;
}

@media only screen and (max-width: 480px) {
  .hero_header__inner {
    padding: 0 16px;
  }
}
