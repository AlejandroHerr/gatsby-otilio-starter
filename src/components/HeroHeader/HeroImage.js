// @flow
/* globals document */
import React, { PureComponent } from 'react';
import Img from 'gatsby-image';

import type { SizeImageSharpType } from '../../types/gatsby';

import styles from './HeroHeader.module.scss';

type PropsType = {
  image: SizeImageSharpType,
};

type StateType = {
  translateY: number,
};

export default class HeroImage extends PureComponent<PropsType, StateType> {
  handleScroll: () => void;

  constructor(props: PropsType) {
    super(props);

    this.state = {
      translateY: 0,
    };

    this.handleScroll = this.handleScroll.bind(this);
  }

  componentDidMount() {
    document.addEventListener('scroll', this.handleScroll);
  }

  componentWillUnmount() {
    document.removeEventListener('scroll', this.handleScroll);
  }

  handleScroll() {
    const scrollTop = (document.documentElement && document.documentElement.scrollTop)
      || 0;

    this.setState({ translateY: scrollTop / 5 });
  }
  render() {
    const { image } = this.props;
    const { translateY } = this.state;

    return (
      <div
        className={styles.hero_image}
        style={{ transform: `translate3d(0px, ${translateY}px, 0px)` }}
      >
        <Img
          sizes={image}
          style={{
            position: 'absolute',
            left: 0,
            top: 0,
            width: '100%',
            height: '100%',
            margin: '0',
          }}
        />
        <div className={styles.hero_image__shadow} />
      </div>
    );
  }
}
