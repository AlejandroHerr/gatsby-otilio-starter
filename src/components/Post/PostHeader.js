import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Link from 'gatsby-link';

import ClockIcon from '../icons/ClockIcon';
import CoverImage from '../CoverImage';

import { imageSharp } from './propTypes';

import styles from './PostHeader.module.scss';

const getDisplayDate = date => date.toLocaleDateString('en-us', { day: 'numeric', month: 'short', year: 'numeric' });

const PostHeader = ({
  author, date, image, title, timeToRead,
}) => {
  const postDate = new Date(date);
  return (
    <header className={classnames(styles.post_header, { [styles['post_header--has_image']]: !!image })}>
      <div className={styles.post_header__inner}>
        <h1 className={styles.post_header__title}>
          {title}
        </h1>
        <span className={styles.post_header__body}>
          <Link className={styles.post_header__author} to="/">{author}</Link>
          {' | '}
          <span><ClockIcon /> {timeToRead}</span>
          {' | '}
          <time dateTime={postDate.toISOString()}>{getDisplayDate(postDate)}</time>
        </span>
        {image && (
        <CoverImage
          className={styles.post_header__image}
          shadowClassName={styles.post_header__image_shadow}
          image={image}
        />
      )}
      </div>
    </header>
  );
};


PostHeader.defaultProps = {
  image: null,
};

PostHeader.propTypes = {
  author: PropTypes.string.isRequired,
  date: PropTypes.string.isRequired,
  image: imageSharp,
  title: PropTypes.string.isRequired,
  timeToRead: PropTypes.number.isRequired,
};

export default PostHeader;
