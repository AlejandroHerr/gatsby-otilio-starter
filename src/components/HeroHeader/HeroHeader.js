// @flow
import React, { type Node } from 'react';
import classnames from 'classnames';

import HeroImage from './HeroImage';
import type { SizeImageSharpType } from '../../types/gatsby';

import styles from './HeroHeader.module.scss';

type PropsType = {
  children: Node,
  className?: string,
  image?: ?SizeImageSharpType,
};


const HeroHeader = ({ children, className, image }: PropsType) => (
  <header className={classnames(styles.hero_header, className, { [styles['hero_header--has_image']]: !!image })}>
    <div className={styles.hero_header__inner}>
      {children}
      {image && (
        <HeroImage
          image={image}
        />
      )}
    </div>
  </header>
);

HeroHeader.defaultProps = {
  className: '',
  image: null,
};

export default HeroHeader;
