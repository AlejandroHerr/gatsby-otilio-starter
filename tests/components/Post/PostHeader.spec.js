import React from 'react';
import { shallow } from 'enzyme';

import HeroHeader from '../../../src/components/HeroHeader';
import PostHeader from '../../../src/components/Post/PostHeader';

const defaultProps = {
  description: 'Test description',
  author: 'author',
  date: '2018-04-30T05:00:00.000Z',
  title: 'Test title',
  timeToRead: 1,
};

const setup = (props = {}) => {
  const postHeader = shallow(<PostHeader {...defaultProps} {...props} />);

  return {
    postHeader,
    time: postHeader.find('time'),
    heroHeader: postHeader.find(HeroHeader),
  };
};

describe('Post', () => {
  describe('PostHeader', () => {
    it('should render a HeroHeader', () => {
      const { postHeader, time, heroHeader } = setup();

      const date = new Date(defaultProps.date);
      const dateTime = date.toISOString();
      const dateText = date.toLocaleDateString('en-us', { day: 'numeric', month: 'short', year: 'numeric' });

      expect(time.prop('dateTime')).toBe(dateTime);
      expect(time.text()).toBe(dateText);
      expect(heroHeader).toHaveLength(1);
      expect(postHeader).toMatchSnapshot();
    });
  });
});
