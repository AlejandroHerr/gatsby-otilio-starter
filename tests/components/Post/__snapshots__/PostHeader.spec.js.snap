// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post PostHeader should render a HeroHeader 1`] = `
<HeroHeader
  className=""
  image={null}
>
  <h1
    className="post_header__title"
  >
    Test title
  </h1>
  <span
    className="post_header__body"
  >
    <GatsbyLink
      className="post_header__author"
      to="/"
    >
      author
    </GatsbyLink>
     | 
    <span>
      <ByteIcon
        bold={true}
        className=""
        icon="clock"
      />
       
      1
    </span>
     | 
    <time
      dateTime="2018-04-30T05:00:00.000Z"
    >
      Apr 30, 2018
    </time>
  </span>
</HeroHeader>
`;
